doc/api/async_hooks.md

```diff
@@ -682,6 +682,193 @@ never be called.
+These methods are synchronous and scoped. The callback function will be executed
+synchronously and it's return value will be returned by `runAndReturn`
+and `exitAndReturn`. The store will only be accessible from within the callback
+function and the asynchronous operations created within this scope.
+If the callback throws an error, `runAndReturn` and `exitAndReturn` will throw
+it too.
+
+These methods provide good scoping while being synchronous.
+
+#### What are the risks of `enterSync/exitSync`
+These methods are to be used only if you know what you are doing. After calling
+`enterSync`, the context will be propagated to everything that follows.
+For instance, in the case of an event emitter, if the first executed listener
+calls `enterSync`, the following ones will have access to the store.
+```js
+emitter.on('event', () => {
+    asyncContext.enterSync();
+});
+emitter.on('event', () => {
+    asyncContext.getStore(); // will return  the store.
+});
+```
+If an instance of AsyncContext is used for error management (for instance with
+`process.setUncaughtExceptionCaptureCallback`). All exceptions will be
+associated with the context.
+
+These methods are less safe than `run/exit` and `runAndReturn/exitAndReturn` as
+they have no scope.
+
+### new AsyncContext()
+
+* Returns: {Object}
+
+Creates a new instance of `AsyncContext`. Storage is only provided inside a `run`
+ method call.
+
+### asyncContext.run(callback[, ...args])
+
+* `callback` {Function}
+  * `store` {Map}
+* `...args` {any}
+
+Calling `asyncContext.run(callback)` will create a new asynchronous context.
+This method will call the callback with a `Map` known as the store as argument.
+The callback will be ran asynchronously.
+
+If an error is thrown by the callback function. It won't ba caught by a
+a `try/catch` block as the callback is ran in a new asynchronous resource.
+
+Optionally, arguments can be passed to the function after the store.
+
+This store will be persistent through the following asynchronous calls.
+
+Currently there is a known limitation with this method. If called within an
+`async function` the context will be lost after the next `await` call.
+
+### asyncContext.exit(callback[, ...args])
+
+* `callback` {Function}
+  * `store` {Map}
+* `...args` {any}
+
+Calling `asyncContext.exit(callback)` will exit the context.
+Asynchronous operations following the callback cannot access the
+ store but its still available for the sync part after `exit()`.
+The callback will be ran asynchronously.
+
+Optionally, arguments can be passed to the function.
+
+If an error is thrown by the callback function. It won't ba caught by a
+a `try/catch` block as the callback is ran in a new asynchronous resource.
+
+Currently there is a known limitation with this method. If called within an
+`async function` the context will be still available after the next `await`
+call.
+
+### asyncContext.runAndReturn(callback[, ...args])
+
+* `callback` {Function}
+  * `store` {Map}
```

Comment:
Not sure if it is helpful to pass `store` here as it requires the user to change the signature of the function passed.
I assume here that it happens frequently that it's just needed to wrap some call with a new context and the easy way would be to just wrap it by `rc = context.runAndReturn(cb, original args)`.
The store can be accessed anyway via `context.getStore()`. And I think it's much saver to use `getStore()` as the callback passed may have further calls to enter/exit or awaits.

Reply:
make sense! I'm changing that now